nshmdb/fault.py

```diff
@@ -0,0 +1,476 @@
+
+
+_KM_TO_M = 1000
+
+
+@dataclasses.dataclass
+class FaultPlane:
+    """A representation of a single plane of a Fault.
+
+    This class represents a single plane of a fault, providing various
+    properties and methods for calculating its dimensions, orientation, and
+    converting coordinates between different reference frames.
+
+    Attributes
+    ----------
+    rake : float
+        The rake angle of the fault plane.
+    """
+
+    _corners: np.ndarray
+    rake: float
+
+    @property
+    def corners(self) -> np.ndarray:
+        """
+
+        Returns
+        -------
+        np.ndarray
+            The corners of the fault plane in (lat, lon, depth) format.
+        """
+        return qcore.coordinates.nztm_to_wgs_depth(self._corners)
+
+    @property
+    def length_m(self) -> float:
+        """
+        Returns
+        -------
+        float
+            The length of the fault plane (in metres).
+        """
+        return np.linalg.norm(self._corners[1] - self._corners[0])
+
+    @property
+    def width_m(self) -> float:
+        """
+        Returns
+        -------
+        float
+            The width of the fault plane (in metres).
+        """
+        return np.linalg.norm(self._corners[-1] - self._corners[0])
+
+    @property
+    def bottom_m(self) -> float:
+        """
+        Returns
+        -------
+        float
+            The bottom depth (in metres).
+        """
+        return self._corners[-1, -1]
+
+    @property
+    def width(self) -> float:
+        """
+        Returns
+        -------
+        float
+            The width of the fault plane (in kilometres).
+        """
+        return self.width_m / _KM_TO_M
+
+    @property
+    def length(self) -> float:
+        """
+        Returns
+        -------
+        float
+            The length of the fault plane (in kilometres).
+        """
+        return self.length_m / _KM_TO_M
+
+    @property
+    def projected_width_m(self) -> float:
+        """
+        Returns
+        -------
+        float
+            The projected width of the fault plane (in metres).
+        """
+        return self.length_m * np.cos(np.radians(self.dip))
+
+    @property
+    def projected_width(self) -> float:
+        """
+        Returns
+        -------
+        float
+            The projected width of the fault plane (in kilometres).
+        """
+        return self.projected_width / _KM_TO_M
+
+    @property
+    def strike(self) -> float:
+        """
+        Returns
+        -------
+        float
+            The bearing of the strike direction of the fault
+            (from north; in degrees)
+        """
+
+        north_direction = np.array([1, 0, 0])
+        up_direction = np.array([0, 0, 1])
+        strike_direction = self._corners[1] - self._corners[0]
+        return qcore.geo.oriented_bearing_wrt_normal(
+            north_direction, strike_direction, up_direction
+        )
+
+    @property
+    def dip_dir(self) -> float:
```

Comment:
Isn't this always strike + 90?

Reply:
Also where do you use this? Just curious since I have never come across it being used

Reply:
Solvis defines DipDir in the json file. Here is an example entry: 

```json
{
      "type": "Feature",
      "id": 0,
      "properties": {
        "FaultID": 0,
        "FaultName": "Acton, Subsection 0",
        "DipDeg": 60.0,
        "Rake": 110.0,
        "LowDepth": 27.12,
        "UpDepth": 0.0,
        "DipDir": 94.3,
        "AseismicSlipFactor": 0.0,
        "CouplingCoeff": 1.0,
        "SlipRate": 1.1875,
        "ParentID": 0,
        "ParentName": "Acton",
        "SlipRateStdDev": 0.50237
      },
      "geometry": {
        "type": "LineString",
        "coordinates": [
            ...
        ]
      }
    }
``` 

It's because the fault segments are not simple rectangles anymore, but instead parallelograms.